odohclient.go

```diff
@@ -0,0 +1,184 @@
+}
+
+// Check the HTTP response status code, parse out the response DNS message and decrypt it.
+func (d *ODoHClient) decodeProxyResponse(resp *http.Response, queryContext odoh.QueryContext) (*dns.Msg, error) {
+	if resp.StatusCode < 200 || resp.StatusCode > 299 {
+		d.proxy.metrics.err.Add(fmt.Sprintf("http%d", resp.StatusCode), 1)
+		return nil, fmt.Errorf("unexpected status code %d", resp.StatusCode)
+	}
+	if resp.Header.Get("content-type") != "application/oblivious-dns-message" {
+		return nil, errors.New("received invalid odoh header from proxy")
+	}
+	rb, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		d.proxy.metrics.err.Add("read", 1)
+		return nil, err
+	}
+	odohQueryResponse, err := odoh.UnmarshalDNSMessage(rb)
+	if err != nil {
+		d.proxy.metrics.err.Add("decode", 1)
+		return nil, err
+	}
+	decryptedResponse, err := queryContext.OpenAnswer(odohQueryResponse)
+	if err != nil {
+		d.proxy.metrics.err.Add("decrypt", 1)
+		return nil, err
+	}
+	a := new(dns.Msg)
+	err = a.Unpack(decryptedResponse)
+	if err != nil {
+		d.proxy.metrics.err.Add("unpack", 1)
+	} else {
+		d.proxy.metrics.response.Add(rCode(a), 1)
+	}
+	return a, err
+}
+
+// Modify a DoH request for proxy-use. The URL and headers need to be updated.
+func (d *ODoHClient) customizeRequest(req *http.Request) {
+	req.Header.Set("content-type", "application/oblivious-dns-message")
+	req.Header.Add("accept", "application/oblivious-dns-message")
+	query := req.URL.Query()
+	query.Add("targethost", d.targetName)
+	query.Add("targetpath", d.targetPath)
+	req.URL.RawQuery = query.Encode()
+}
+
+// Marshal and encrypt the original query with the key for the target.
+func (d *ODoHClient) buildTargetQuery(q *dns.Msg) (odoh.ObliviousDNSMessage, odoh.QueryContext, error) {
+	config, err := d.getTargetConfig()
+	if err != nil {
+		return odoh.ObliviousDNSMessage{}, odoh.QueryContext{}, err
+	}
+	key := config.Contents
+	msg, err := q.Pack()
+	if err != nil {
+		return odoh.ObliviousDNSMessage{}, odoh.QueryContext{}, err
+	}
+	odohQ := odoh.CreateObliviousDNSQuery(msg, 0)
+	return key.EncryptQuery(odohQ)
+}
+
+// Get the current (cached) target config or refresh it if expired.
+func (d *ODoHClient) getTargetConfig() (*odoh.ObliviousDoHConfig, error) {
+	var err error
+	d.mu.Lock()
+	defer d.mu.Unlock()
+	if d.odohConfig == nil || time.Now().After(d.odohConfigExpiry) {
+		d.odohConfig, d.odohConfigExpiry, err = d.refreshTargetKey()
+	}
+	return d.odohConfig, err
+}
+
+// Load the key by making a DoH query to the proxy, then cache it.
+func (d *ODoHClient) refreshTargetKey() (*odoh.ObliviousDoHConfig, time.Time, error) {
+	query := new(dns.Msg)
+	query.SetQuestion(dns.Fqdn(d.targetName), dns.TypeHTTPS)
+	response, err := d.proxy.Resolve(query, ClientInfo{})
+	if err != nil {
+		return nil, time.Time{}, err
+	}
+
+	for _, answer := range response.Answer {
+		https, ok := answer.(*dns.HTTPS)
+		if !ok {
+			continue
+		}
+		for _, value := range https.Value {
+			if value.Key() != 32769 {
```

Comment:
unfortunately, not all odoh servers publish https/svcb record. in fact, [RFC-9230](https://datatracker.ietf.org/doc/rfc9230/) has removed it from spec (from `draft-07` to be precise).

attempting to fetch odoh-config from both, `/.well-known/odohconfigs` and https/svcb records, is a safer bet. we do this for our android app, rethink ([code](https://github.com/celzero/firestack/blob/fb23a0fd4ff301701bf4d27d357c76473b9430f1/intra/doh/odoh.go#L142-L158)). cloudflare's reference implementation also does the same.